scheds/rust/scx_lavd/src/bpf/main.bpf.c

```diff
@@ -2321,69 +2677,150 @@ s32 BPF_STRUCT_OPS(lavd_init_task, struct task_struct *p,
+	if (cpumask)
+		bpf_cpumask_release(cpumask);
+
+	return 0;
+}
+
+static int init_cpumasks(void)
+{
+	struct bpf_cpumask *active;
+	int err = 0;
+	u32 cpu;
+
+	bpf_rcu_read_lock();
+	err = calloc_cpumask(&active_cpumask);
+	active = active_cpumask;
+	if (err || !active)
+		goto out;
+
+	err = calloc_cpumask(&ovrflw_cpumask);
+	if (err)
+		goto out;
 
 	/*
-	 * Create a central task queue.
+	 * Initially activate all CPUs until we know the system load.
 	 */
-	err = scx_bpf_create_dsq(LAVD_GLOBAL_DSQ, -1);
-	if (err) {
-		scx_bpf_error("Failed to create a shared DSQ");
-		return err;
+	bpf_for(cpu, 0, nr_cpus_onln) {
+		bpf_cpumask_set_cpu(cpu, active);
 	}
 
-	/*
-	 * Initialize per-CPU context
-	 */
+out:
+	bpf_rcu_read_unlock();
+	return err;
+}
+
+static s32 init_per_cpu_ctx(void)
+{
+	int cpu;
+	int err;
+
 	bpf_for(cpu, 0, nr_cpus_onln) {
 		struct cpu_ctx *cpuc = get_cpu_ctx_id(cpu);
 		if (!cpuc) {
 			scx_bpf_error("Failed to lookup cpu_ctx: %d", cpu);
 			return -ESRCH;
 		}
+
+		err = calloc_cpumask(&cpuc->tmp_a_mask);
+		if (err)
+			return err;
+
+		err = calloc_cpumask(&cpuc->tmp_o_mask);
+		if (err)
+			return err;
+
 		cpu_ctx_init_online(cpuc, cpu);
 	}
 
-	/*
-	 * Initialize the last update clock and the update timer to track
-	 * system-wide CPU load.
-	 */
-	memset(__sys_cpu_util, 0, sizeof(__sys_cpu_util));
+	return 0;
+}
+
+static s32 init_sys_stat(void)
+{
+	struct bpf_timer *timer;
+	u64 now;
+	u32 key = 0;
+	int err;
+
+	memset(__sys_stats, 0, sizeof(__sys_stats));
 	now = bpf_ktime_get_ns();
-	__sys_cpu_util[0].last_update_clk = now;
-	__sys_cpu_util[1].last_update_clk = now;
+	__sys_stats[0].last_update_clk = now;
+	__sys_stats[1].last_update_clk = now;
+	__sys_stats[0].nr_active = nr_cpus_onln;
+	__sys_stats[1].nr_active = nr_cpus_onln;
 
 	timer = bpf_map_lookup_elem(&update_timer, &key);
 	if (!timer) {
 		scx_bpf_error("Failed to lookup update timer");
 		return -ESRCH;
 	}
 	bpf_timer_init(timer, &update_timer, CLOCK_BOOTTIME);
-	bpf_timer_set_callback(timer, update_timer_fn);
+	bpf_timer_set_callback(timer, update_timer_cb);
 	err = bpf_timer_start(timer, LAVD_CPU_UTIL_INTERVAL_NS, 0);
 	if (err) {
 		scx_bpf_error("Failed to arm update timer");
 		return err;
 	}
 
+	return 0;
+}
+
+s32 BPF_STRUCT_OPS_SLEEPABLE(lavd_init)
+{
+	int err;
+
+	/*
+	 * Create a central task queue.
+	 */
+	err = scx_bpf_create_dsq(LAVD_GLOBAL_DSQ, -1);
+	if (err) {
+		scx_bpf_error("Failed to create a shared DSQ");
+		return err;
+	}
+
+	/*
+	 * Initialize per-CPU context.
+	 */
+	err = init_per_cpu_ctx();
+	if (err)
+		return err;
+
+	/*
+	 * Initialize the last update clock and the update timer to track
+	 * system-wide CPU load.
+	 */
+	err = init_sys_stat();
+	if (err)
+		return err;
+
+	/*
+	 * Allocate cpumask for task compaction.
+	 *  - active CPUs: a group of CPUs will be used for now.
+	 *  - overflow CPUs: a pair of hyper-twin which will be used when there
+	 *    is no idle active CPUs.
+	 */
+	init_cpumasks();
```

Comment:
Should we check the return value here as well?

Reply:
Oh, you are right. Fixed.